src/Builder/Select/QueryExpressionVisitor.php

```diff
@@ -0,0 +1,178 @@
+                    return sprintf('%s:(%s)', $field, $this->valueToString($value, ' OR ', '"'));
+                }
+
+                return sprintf('%s:%s', $field, $this->valueToString($value, ',', '"'));
+            case Comparison::LIKE:
+            case Comparison::MATCH:
+                if (\is_array($value)) {
+                    return sprintf('%s:(%s)', $field, $this->valueToString($value, ' OR ', '', false));
+                }
+
+                return sprintf('%s:%s', $field, $this->valueToString($value, ',', '', false));
+            case Comparison::NIN:
+                if (\is_array($value)) {
+                    return sprintf('-%s:(%s)', $field, $this->valueToString($value, ' OR ', '"'));
+                }
+
+                return sprintf('-%s:%s', $field, $this->valueToString($value, ',', '"'));
+            case Comparison::REGEXP:
+                if ('/' !== $value[0]) {
+                    $value = sprintf('/%s/', $value);
+                }
+
+                return sprintf('%s:%s', $field, $this->valueToString($value, ',', '', false));
+            default:
+                throw new RuntimeException('Unknown comparison operator: '.$expression->getOperator());
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function walkValue(Value $value)
+    {
+        return $value->getValue();
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws \Solarium\Exception\RuntimeException
+     */
+    public function walkCompositeExpression(ExpressionInterface $expr)
+    {
+        $comparisons = [];
+
+        foreach ($expr->getComparisons() as $child) {
+            $comparisons[] = $this->dispatch($child);
+        }
+
+        switch ($expr->getType()) {
+            case CompositeComparison::TYPE_AND:
+                return implode(' AND ', $comparisons);
+            case CompositeComparison::TYPE_OR:
+                return implode(' OR ', $comparisons);
+            default:
+                throw new RuntimeException('Unknown composite '.$expr->getType());
+        }
+    }
+
+    /**
+     * @param mixed  $value
+     * @param string $separator
+     * @param string $quote
+     * @param bool   $escape
+     *
+     * @return string
+     */
+    private function valueToString($value, string $separator = ',', string $quote = '', bool $escape = true): string
+    {
+        if (\is_array($value)) {
+            $ret = [];
+
+            foreach ($value as $v) {
+                $ret[] = $this->typedValueToString($v, $quote, $escape);
+            }
+
+            return implode($separator, $ret);
+        }
+
+        return $this->typedValueToString($value, $quote, $escape);
+    }
+
+    /**
+     * @param mixed  $value
+     * @param string $quote
+     * @param bool   $escape
+     *
+     * @return string
+     */
+    private function typedValueToString($value, string $quote = '', $escape = true): string
+    {
+        if (null === $value) {
+            return '[* TO *]';
```

Comment:
This will fail on location data. Use Helper::rangeQuery().

Reply:
looks like the helper returns the same for an empty value?
https://github.com/solariumphp/solarium/blob/d6d1adb67a2cc905a3783bf57cec152d495e6fcc/src/Core/Query/Helper.php#L183-L196